File: go/code_smells/couplers/feature_envy.go
package couplers

import "fmt"

// Customer represents a customer
type Customer struct {
	Name    string
	Email   string
	Address string
	Phone   string
}

// Invoice represents an invoice
type Invoice struct {
	InvoiceNumber string
	Amount        float64
	Customer      Customer
}

// InvoiceService demonstrates feature envy code smell
// The method is more interested in the Customer class than its own class
type InvoiceService struct{}

// SendInvoice sends an invoice to a customer
// This method is "envious" of Customer's data
func (is *InvoiceService) SendInvoice(invoice Invoice) error {
	// This method knows too much about Customer's internal structure
	// and uses Customer's data more than its own

	// Formatting customer info
	customerInfo := fmt.Sprintf(
		"Customer: %s\nEmail: %s\nAddress: %s\nPhone: %s",
		invoice.Customer.Name,
		invoice.Customer.Email,
		invoice.Customer.Address,
		invoice.Customer.Phone,
	)

	// Validating customer data
	if invoice.Customer.Email == "" {
		return fmt.Errorf("customer email is required")
	}
	if invoice.Customer.Address == "" {
		return fmt.Errorf("customer address is required")
	}

	fmt.Printf("Sending invoice %s for %.2f to:\n%s\n",
		invoice.InvoiceNumber, invoice.Amount, customerInfo)

	return nil
}

// Exercise: Refactor this code to eliminate feature envy
// 1. Move the customer-related logic to the Customer struct
// 2. Create methods like Customer.FormatInfo() and Customer.Validate()
// 3. Keep the InvoiceService focused on invoice-related operations

File: java/src/main/java/com/refactoring/codesmells/couplers/FeatureEnvy.java
package com.refactoring.codesmells.couplers;

public class FeatureEnvy {

    public static class Customer {
        private final String name;
        private final String email;
        private final String address;
        private final String phone;

        public Customer(String name, String email, String address, String phone) {
            this.name = name;
            this.email = email;
            this.address = address;
            this.phone = phone;
        }

        public String getName() {
            return name;
        }

        public String getEmail() {
            return email;
        }

        public String getAddress() {
            return address;
        }

        public String getPhone() {
            return phone;
        }
    }

    public static class Invoice {
        private final String invoiceNumber;
        private final double amount;
        private final Customer customer;

        public Invoice(String invoiceNumber, double amount, Customer customer) {
            this.invoiceNumber = invoiceNumber;
            this.amount = amount;
            this.customer = customer;
        }

        public String getInvoiceNumber() {
            return invoiceNumber;
        }

        public double getAmount() {
            return amount;
        }

        public Customer getCustomer() {
            return customer;
        }
    }

    public static class InvoiceService {
        // This method is "envious" of Customer's data
        public void sendInvoice(Invoice invoice) {
            // This method knows too much about Customer's internal structure
            // and uses Customer's data more than its own

            // Formatting customer info
            String customerInfo = String.format(
                    "Customer: %s%nEmail: %s%nAddress: %s%nPhone: %s",
                    invoice.getCustomer().getName(),
                    invoice.getCustomer().getEmail(),
                    invoice.getCustomer().getAddress(),
                    invoice.getCustomer().getPhone()
            );

            // Validating customer data
            if (invoice.getCustomer().getEmail() == null || invoice.getCustomer().getEmail().isEmpty()) {
                throw new IllegalArgumentException("customer email is required");
            }
            if (invoice.getCustomer().getAddress() == null || invoice.getCustomer().getAddress().isEmpty()) {
                throw new IllegalArgumentException("customer address is required");
            }

            System.out.printf("Sending invoice %s for %.2f to:%n%s%n",
                    invoice.getInvoiceNumber(), invoice.getAmount(), customerInfo);
        }
    }
}
